src/cryptonote_core/service_node_list.cpp

```diff
@@ -4941,6 +4953,65 @@ static std::string serialize_snl_directly(Archive& ar, service_node_list::state_
     return result;
 }
 
+namespace {
+    struct zstd_decomp_freer {
+        void operator()(ZSTD_DStream* z) const { ZSTD_freeDStream(z); }
+    };
+
+    using zstd_decomp_ptr = std::unique_ptr<ZSTD_DStream, zstd_decomp_freer>;
+
+    std::string zstd_compress(std::string_view data, int level, std::string_view prefix = {}) {
+        std::string compressed;
+        compressed.resize(ZSTD_compressBound(data.size()) + prefix.size());
+        auto size = ZSTD_compress(
+                compressed.data() + prefix.size(),
+                compressed.size() - prefix.size(),
+                data.data(),
+                data.size(),
+                level);
+        if (ZSTD_isError(size))
+            throw std::runtime_error{"Compression failed: {}"_format(ZSTD_getErrorName(size))};
+
+        compressed.resize(prefix.size() + size);
+        std::copy(prefix.begin(), prefix.end(), compressed.begin());
+        return compressed;
+    }
+
+    std::optional<std::string> zstd_decompress(std::string_view data) {
+        zstd_decomp_ptr z_decompressor{ZSTD_createDStream()};
+        auto* zds = z_decompressor.get();
+
+        ZSTD_initDStream(zds);
+        ZSTD_inBuffer input{.src = data.data(), .size = data.size(), .pos = 0};
+        std::array<char, 4096> out_buf;
+        ZSTD_outBuffer output{.dst = out_buf.data(), .size = out_buf.size(), .pos = 0};
+
+        std::string decompressed;
```

Comment:
You can query the decompressed size using `ZSTD_getFrameContentSize`

```cpp
uint64_t decompressed_size = ZSTD_getFrameContentSize(data.data(), data.size());
assert(decompressed_size != ZSTD_CONTENTSIZE_UNKNOWN && decompressed_size != ZSTD_CONTENTSIZE_ERROR);
```

Note that the decompressed size is always available if your use `ZSTD_compress` so in practice `ZSTD_CONTENSIZE_UNKNOWN` is not possible. `ZSTD_CONTENTSIZE_ERROR` can happen if the blob was corrupted so that one can be handled by returning `std::nullopt`.

Reply:
Ah, thanks for that.  I've updated it with that to properly reserve now.